dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java

```diff
@@ -457,57 +450,56 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
 			}
             if (bundle == null)
             {
                 log.trace("Creating bundle in item.");
-                dspaceBitstream = dspaceItem.createSingleBitstream(inputStream);
+                dspaceBitstream = itemService.createSingleBitstream(context, inputStream, dspaceItem);
             }
             else
             {
                 log.trace("Getting bundle from item.");
-                dspaceBitstream = bundle.createBitstream(inputStream);
+                dspaceBitstream = bitstreamService.create(context, bundle, inputStream);
             }
 
-            dspaceBitstream.setSource("DSpace Rest api");
+            dspaceBitstream.setSource(context, "DSpace REST API");
 
             // Set bitstream name and description
             if (name != null)
             {
                 if (BitstreamResource.getMimeType(name) == null)
                 {
-                    dspaceBitstream.setFormat(BitstreamFormat.findUnknown(context));
+                    dspaceBitstream.setFormat(context, bitstreamFormatService.findUnknown(context));
                 }
                 else
                 {
-                    dspaceBitstream.setFormat(BitstreamFormat.findByMIMEType(context, BitstreamResource.getMimeType(name)));
+                    bitstreamService.setFormat(context, dspaceBitstream, bitstreamFormatService.findByMIMEType(context, BitstreamResource.getMimeType(name)));
                 }
-                dspaceBitstream.setName(name);
+
+                dspaceBitstream.setName(context, name);
             }
             if (description != null)
             {
-                dspaceBitstream.setDescription(description);
+                dspaceBitstream.setDescription(context, description);
             }
 
-            dspaceBitstream.update();
-
             // Create policy for bitstream
             if (groupId != null)
             {
-                bundles = dspaceBitstream.getBundles();
+                List<BundleBitstream> bundleBitstreams = dspaceBitstream.getBundles();
                 for (Bundle dspaceBundle : bundles)
                 {
-                    List<org.dspace.authorize.ResourcePolicy> bitstreamsPolicies = dspaceBundle.getBitstreamPolicies();
+                    List<org.dspace.authorize.ResourcePolicy> bitstreamsPolicies = bundleService.getBitstreamPolicies(context, dspaceBundle);
 
                     // Remove default bitstream policies
                     List<org.dspace.authorize.ResourcePolicy> policiesToRemove = new ArrayList<org.dspace.authorize.ResourcePolicy>();
-                    for (org.dspace.authorize.ResourcePolicy policy : bitstreamsPolicies)
-                    {
-                        if (policy.getResourceID() == dspaceBitstream.getID())
+                    for (org.dspace.authorize.ResourcePolicy policy : bitstreamsPolicies) {
+                        if (policy.getdSpaceObject() == dspaceBitstream)
```

Comment:
This type of compare won't work as you are comparing objects & not primitive types.